akka-remote/src/main/scala/akka/remote/artery/MetadataEnvelopeSerializer.scala

```diff
@@ -66,14 +77,18 @@ private[akka] final class MetadataMapRendering(val metadataMap: MetadataMap[Byte
   import MetadataEnvelopeSerializer._
 
   def render(): ByteString =
-    if (metadataMap.isEmpty) ByteString.empty
-    else {
+    if (metadataMap.isEmpty) {
+      // usually no-one will want to render an empty metadata section - it should not be there at all
+      EmptyRendered
+    } else {
       implicit val _ByteOrder = ByteOrder.LITTLE_ENDIAN
 
-      val size = metadataMap.usedSlots * 4 /* bytes (Int) */ + metadataMap.foldLeftValues(0)(_ + _.length)
-      val b = new ByteStringBuilder
-      b.sizeHint(size)
+      // TODO optimise this, we could just count along the way and then prefix with the length
+      val totalSize = 4 /* length int field */ + metadataMap.usedSlots * 4 /* metadata length */ + metadataMap.foldLeftValues(0)(_ + _.length)
```

Comment:
this is not nice, can be done differently (counting the length while we write things, and then prepending with it)